src/gennylib/include/gennylib/topology.hpp

```diff
@@ -0,0 +1,219 @@
+// Copyright 2021-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#ifndef HEADER_6D7AD2AF_7CD1_49E4_9712_741ABD354DFC_INCLUDED
+#define HEADER_6D7AD2AF_7CD1_49E4_9712_741ABD354DFC_INCLUDED
+
+#include <vector>
+#include <variant>
+#include <mutex>
+#include <sstream>
+
+#include <gennylib/v1/PoolFactory.hpp>
+#include <gennylib/service.hpp>
+
+namespace genny {
+
+class AbstractTopologyDescription;
+class MongodDescription;
+class MongosDescription;
+class ReplSetDescription;
+class ShardedDescription;
+
+/**
+ * Inherit from TopologyVisitor and override particular visit methods
+ * to implement an algorithm that operates on a cluster. Pass the visitor
+ * to a Topology object to execute.
+ *
+ * The idea is to create a visitor that focuses on each part of the cluster,
+ * allow the Topology object to handle traversal, and keep application-level
+ * code as topology-agnostic as possible.
+ */
+class TopologyVisitor {
+public:
+    virtual void visitTopLevelPre(const AbstractTopologyDescription&) {}
+    virtual void visitTopLevelPost(const AbstractTopologyDescription&) {}
+
+    virtual void visitMongodDescription(const MongodDescription&) {}
+    virtual void visitMongosDescription(const MongosDescription&) {}
+
+    virtual void visitReplSetDescriptionPre(const ReplSetDescription&) {}
+    virtual void visitReplSetDescriptionPost(const ReplSetDescription&) {}
+
+    virtual void visitShardedDescriptionPre(const ShardedDescription&) {}
+    virtual void visitShardedDescriptionPost(const ShardedDescription&) {}
+
+
+    /** Misc hooks that most visitors won't need. **/
+
+    // Called between mongods in a replica set.
+    virtual void visitMongodDescriptionBetween(const ReplSetDescription&) {}
+
+    // Called before/after/between visiting shards.
+    virtual void visitShardsPre(const ShardedDescription&) {}
+    virtual void visitShardsPost(const ShardedDescription&) {}
+    virtual void visitShardsBetween(const ShardedDescription&) {}
+
+    // Called before/after/between visiting mongoses.
+    virtual void visitMongosesPre(const ShardedDescription&) {}
+    virtual void visitMongosesPost(const ShardedDescription&) {}
+    virtual void visitMongosesBetween(const ShardedDescription&) {}
+
+    virtual ~TopologyVisitor() {}
```

Comment:
let's make this `=0` to make this a pure virtual class.

Reply:
I thought pure virtual destructors were bad practice? Because the linker requires a definition for the destructor whether it's pure virtual or not, since every destructor call also calls the parent's destructor, its parent's, etc so a definition would need to be added outside the class declaration and having at least one function be pure virtual is all that's needed to make the class an abstract class?